cpp/src/sqls/sqls_rtti_comp.h

```diff
@@ -127,198 +260,173 @@ struct LesserRTTI
+      ColType res1 = LesserRTTI::at<ColType>(col_index, row1, columns);
+      ColType res2 = LesserRTTI::at<ColType>(col_index, row2, columns);
       
       if( res1 != res2 )
         return State::False;
       else
         return State::Undecided;
     }
-    
-  private:
-    IndexT row1_;
-    IndexT row2_;
   };
 
   struct OpEqualV
   {
-    __device__
-    OpEqualV(const void* const * vals, IndexT row):
-      target_vals_(vals),
-      row_(row)
-    {
-    }
-    
     template<typename ColType>
     __device__
-    State operator() (int col_index,
-		      const void* const * columns,
-		      ColType )
+    State operator() (const void* const * vals,
+                      IndexT row,
+                      int col_index,
+                      const void* const * columns)
     {
-      ColType res1 = LesserRTTI::at<ColType>(col_index, row_, columns);
-      ColType res2 = LesserRTTI::at<ColType>(col_index, 0, target_vals_);
+      ColType res1 = LesserRTTI::at<ColType>(col_index, row, columns);
+      ColType res2 = LesserRTTI::at<ColType>(col_index, 0, vals);
       
       if( res1 != res2 )
         return State::False;
       else
         return State::Undecided;
     }
-
-  private:
-    const void* const * target_vals_;
-    IndexT row_;
   };
 
-  template<typename Predicate>
-  __device__
-  State type_dispatcher(Predicate pred, gdf_dtype col_type, int col_index) const
-  {
-    switch( col_type )
-    {
-      case GDF_INT8:
-      {
-        using ColType = int8_t;//char;
-        
-        ColType dummy=0;
-        return pred(col_index, columns_, dummy);
-      }
-      case GDF_INT16:
-      {
-        using ColType = int16_t;//short;
-
-        ColType dummy=0;
-        return pred(col_index, columns_, dummy);
-      }
-      case GDF_INT32:
-      {
-        using ColType = int32_t;//int;
-
-        ColType dummy=0;
-        return pred(col_index, columns_, dummy);
-      }
-      case GDF_INT64:
-      {
-        using ColType = int64_t;//long;
-
-        ColType dummy=0;
-        return pred(col_index, columns_, dummy);
-      }
-      case GDF_FLOAT32:
-      {
-        using ColType = float;
-
-        ColType dummy=0;
-        return pred(col_index, columns_, dummy);
-      }
-      case GDF_FLOAT64:
-      {
-        using ColType = double;
-
-        ColType dummy=0;
-        return pred(col_index, columns_, dummy);
-      }
-
-      default:
-	      assert( false );//type not handled
-    }
-    return State::Undecided;
-  }
-  
-  const void* const * columns_;
-  const int* const rtti_;
+  const void* const * columns_{nullptr};
+  const gdf_valid_type* const * valids_{nullptr};
+  const int* const rtti_{nullptr};
   size_t sz_;
-  const void* const * vals_; //for filtering
+  const void* const * vals_{nullptr}; //for filtering
+  int8_t* const asc_desc_flags_{nullptr}; //array of 0 and 1 that allows us to know whether or not a column should be sorted ascending or descending

```

Comment:
The values of the flags themselves should also be `const`. Better chance that these values will be cached if marked as `const`, and seeing as though you access them for every column, for every comparison, you want to make sure they're cached. 
```suggestion
  int8_t const * const asc_desc_flags_{nullptr}; //array of 0 and 1 that allows us to know whether or not a column should be sorted ascending or descending
```